docs/parts/options.md
## Options

Options for DeviantArt filtered are configured through the [Management Panel/Screen](#opening-the-management-panelscreen) on the [Options](#options-page) tab.

### Use Placeholders for Filtered Deviations

`When enabled, a placeholder image is displayed instead of the actual thumbnail for filtered deviations. Disable this setting to completely remove filtered deviation thumbnails.`

If you want to see/know when a deviation was filtered (and why), enable this option (it is enabled by default). This will show a placeholder instead of the deviation's thumbnail (similar to Mature Content that is blocked). If you would rather just completely hide deviations matching any of your filters, disable this option, and the thumbnails will be removed entirely.

### Background/Logo/Text Color for Placeholders

`Set a custom color for the background/logo/text of the placeholder shown in place of filtered deviations.`

Use these three options to control the style of the placeholders that are inserted for filtered deviations. These settings are ignored when placeholders are disabled altogether.

### Management Panel Type

`Specifies where/how the main management panel is displayed when clicking the DeviantArt Filter logo by the address bar.`

This option allows you to control if the DeviantArt Filter Management page is opened in a new tab in your current browser window, or if a new popup window is opened. The default behavior for this option is to use a new tab in the current window.

### Metadata Batch Size

`Specifies the number of deviations for which to load metadata at a time. Setting this value to a lower number may improve browsing performance.`

This option limits how much metadata (used for category and tag filters) is loaded at once. The default is 96, which means that metadata for up to 96 deviations is retrieved for deviations on the page and loaded for the tag and category filters to use. Lowering this number may improve browsing performance, so you can try different values for this option to see if you notice any improvements.

### Number of Days to Cache Metadata

`Specifies the number of days to cache metadata (for tags and categories) locally for deviations. Set this value to disable local caching completely.`

By default, DeviantArt Filter caches the metadata for deviations while browsing [DeviantArt](https://www.deviantart.com) on your computer. This is done so that your category and tag filters can be applied more quickly on subsequent visits, as retrieving metadata for each page of deviations takes 2-5 seconds on an average internet connection. You can disable the local cache completely by setting this option to `0`, but this will cause a noticeable delay before tag and category filters are applied. You can also increase this setting, although setting a value that is too large may cause performance issues.

### Show Metadata Debug Indicators

`When enabled, deviation thumbnails with missing metadata (for tags and categories) have a red outline, and deviations with metadata have a green outline. This is primarily for debugging, but can be a neat visual for curious people.`

This setting, as the description states, is primarily for troubleshooting and debugging. It is used simply to show which deviations have loaded metadata (i.e. their tags and which category they are in). Because metadata has to be fetched from DeviantArt's API, your tag and category filters cannot be applied until that metadata is loaded. The outlines (green for loaded, red for NOT loaded), can help you see why your filters may not seem to be working.


CHANGELOG.md
# DeviantArt Filter Changelog

## Version 5.1 (February 6, 2019)

### New Features

- **Custom Placeholder Styles**. You can now control the colors used for the placeholders on filtered deviations. Management Panel/Screen, go to the Options tab, and set the Background, Logo, and/or Text Colors for Placeholders. Note that these settings are ineffective if you have placeholders disabled.
- **Metadata Batch Size**. A new (somewhat experimental) feature has been implemented to allow control over how much metadata is loaded at once (for category and tag filters). Setting this to a lower value may help your browsing performance.

### Bug Fixes

- Clarify and restrict the use of the hashtag symbol (`#`) when creating tag filters ([#57](https://github.com/rthaut/deviantART-Filter/issues/57))

## Version 5.0.3 (June 22, 2018)

### Bug Fixes

- [Tag and Category Filters do not Work](https://github.com/rthaut/deviantART-Filter/issues/48)
- [Cannot Quick Filter Users from Journal Thumbnails](https://github.com/rthaut/deviantART-Filter/issues/52)

### Release Notes

Upgrading to Version 5.0.3 will reset your cached metadata. **This shouldn't have any noticeable impact, but if do encounter any issues (like metadata not loading, or tag and category filters not working), you may need to restart your browser.**

Also, please be aware that the changes/fixes made in Version 5.0.3 (just like 5.0.1 and 5.0.2) are **primarily for people who are in the Beta Test program for DeviantArt**, which is only available for Core members with paid subscriptions. If you are in the Beta Test program and encounter issues with DeviantArt Filter, please either [send me (rthaut) a private Note on DeviantArt](https://www.deviantart.com/notifications/notes/#to=rthaut) or [create a new issue on GitHub](https://github.com/rthaut/deviantART-Filter/issues) so I can (try to) make DeviantArt Filter compatible with the beta changes.

## Version 5.0.2 (June 16, 2018)

### Bug Fixes

- [User Quick Filter No Longer Works](https://github.com/rthaut/deviantART-Filter/issues/51)

## Version 5.0.1 (June 13, 2018)

### Bug Fixes

- [User Quick Filter Causes All Thumbnails to be Filtered](https://github.com/rthaut/deviantART-Filter/issues/46)
- Fixes the rendering of the category sub-levels when creating category filters

## Version 5.0 (May 13, 2018)

**Version 5 is a complete re-write.** There is a significant amount of new functionality, which, in combination with modern best/common practices, has resulted in a very different user experience for managing your filters. Please see the [section on Opening the Management Panel/Screen](https://github.com/rthaut/deviantART-Filter#opening-the-management-panelscreen) for help.

### New Features

- Filter Tags and Categories
  - In addition to filtering deviations from specific users, you can now filter uses with specific tags and/or submitted to specific categories.
  - Note that there is a slight delay while browsing before your tag and category filters are applied. If your tag and/or category filters do not seem to be working, you may want to [enable the debug indicators for metadata](https://github.com/rthaut/deviantART-Filter#show-metadata-debug-indicators) to see if metadata is actually being loaded while you browse.
- Sort and Page through Filters
  - You can now sort your filters and page through them (instead of scrolling through one giant list).
- Improved Import and Export
  - Filters are exported to a file (instead of requiring you to manually copy/paste the data).
  - The import/export interface has been overhauled to provide detailed information about import results.
  - Drag and drop support for importing filters from a file
- Placeholders Show Why a Thumbnail is Filtered
  - Since you are now able to filter by tag, category, or user, the placeholder image (when placeholders are enabled) shows which filter is applied.
- Support for Translations/Internationalization
  - Currently, DeviantArt Filter is only available in English, but version 5 fully supports translations. **If you want to help translate DeviantArt Filter into other languages, please see [CONTRIBUTING](https://github.com/rthaut/deviantART-Filter/blob/master/CONTRIBUTING.md).**

### Bug Fixes

- [Filters do not Work on Subdomains](https://github.com/rthaut/deviantART-Filter/issues/26)
- [Apply Filtering to Thumbnails in Comments](https://github.com/rthaut/deviantART-Filter/issues/25)


src/scripts/background/options.js
const Options = (() => {

    const Options = {

        /**
         *
         */
        'initDefaultOptions': async function () {
            console.log('[Background] Options.initDefaultOptions()');

            //TODO: this list should probably be a constant that is somehow tied to the "options" data in getOptions()
            var defaults = {
                'managementPanelType': 'tab',
                'metadataBatchSize': 96,
                'metadataCacheTTL': 7,
                'metadataDebug': false,
                'placeholders': true,
                'placeholderBGColor': '#DDE6DA',
                'placeholderLogoColor': '#B4C0B0',
                'placeholderTextColor': '#B4C0B0',
                'privateStorage': 'read'
            };

            const data = await browser.storage.sync.get();

            var options = {};
            for (var option in defaults) {
                if (data[option] === undefined) {
                    options[option] = defaults[option];
                }
            }

            return browser.storage.sync.set(options);
        },

        /**
         *
         * @param {string} option the key of the option
         * @param {string} value the value of the option
         */
        'setOption': function (option, value) {
            console.log('[Background] Options.setOption()', option, value);

            var opt = {};
            opt[option] = value;
            return browser.storage.sync.set(opt);
        },

        /**
         *
         */
        'getOptions': async function () {
            console.log('[Background] Options.getOptions()');

            //TODO: this list (and the properties below) should probably be a constant that is somehow tied to the "defaults"
            // NOTE: the order used below is the display order on the options tab of the management panel
            const data = await browser.storage.sync.get([
                'placeholders',
                'placeholderBGColor',
                'placeholderLogoColor',
                'placeholderTextColor',
                'managementPanelType',
                'metadataBatchSize',
                'metadataCacheTTL',
                'metadataDebug',
            ]);

            var options = [];
            var opt;
            for (var option in data) {
                opt = {
                    'id': option,
                    'name': browser.i18n.getMessage(`Option${option}Name`),
                    'description': browser.i18n.getMessage(`Option${option}Description`),
                    'value': data[option]
                };
                switch (option) {
                    case 'managementPanelType':
                        opt.type = 'radio';
                        opt.values = [
                            {
                                'value': 'window',
                                'label': browser.i18n.getMessage('OptionManagementPanelTypeValueWindow')
                            },
                            {
                                'value': 'tab',
                                'label': browser.i18n.getMessage('OptionManagementPanelTypeValueTab')
                            }
                        ];
                        break;

                    case 'metadataBatchSize':
                        opt.type = 'range';
                        opt.min = 16;
                        opt.max = 96;
                        opt.step = 16;
                        break;

                    case 'metadataCacheTTL':
                        opt.type = 'number';
                        opt.minimum = 0;
                        opt.maximum = 365;
                        break;

                    case 'metadataDebug':
                    case 'placeholders':
                        opt.type = 'checkbox';
                        break;

                    case 'placeholderBGColor':
                    case 'placeholderLogoColor':
                    case 'placeholderTextColor':
                        opt.type = 'color';
                        break;

                    default:
                        opt.type = 'string';
                        break;
                }
                options.push(opt);
            }

            console.log('[Background] Options.getOptions() :: Return', options);

            return { 'options': options };
        }

    };

    return Options;

})();

export default Options;
